File: python/completed/euler0044.py
###
#
# Created by Holden on 5/25/2017
#
# COMPLETED on 6/7/2017
#
# Problem 44 - Pentagon Numbers
#
###

import math


def pentagon_numbers():
    pentagons = [1, 5]
    n = 2

    while True:
        n += 1
        p2 = (n * (3 * n - 1)) // 2
        pentagons.append(p2)
        for p1 in pentagons:
            if (p2 - p1) in pentagons:
                if quadratic_int(3, -1, -2*(p2+p1)):
                    print(p1, p2)
                    return p2 - p1


def quadratic_int(a, b, c):
    d = (b * b) - (4 * a * c)
    if d < 0:
        return False

    if d == 0:
        return (-b / (2 * a)).is_integer()

    x1 = ((-b) + math.sqrt(d)) / (2 * a)
    x2 = ((-b) - math.sqrt(d)) / (2 * a)

    #print(x1, x2)

    return (x1.is_integer() and x1 >= 0) or (x2.is_integer() and x2 >= 0)

print(pentagon_numbers())

File: javascript/completed/euler0044.js
function isPentagonal(p, pentagonalNums) {
  if (pentagonalNums.get(p) === undefined) {
    const n = (Math.sqrt(24 * p + 1) + 1) / 6;
    pentagonalNums.set(p, Number.isInteger(n));
  }
  return pentagonalNums.get(p);
}

function findPentagonal(n) {
  return (n * (3 * n - 1)) / 2;
}

export function solve44() {
  let pentagonalNums = new Map();
  let n = 2;
  let d = 0;
  let p1 = 1;
  let p2 = 5;
  pentagonalNums.set(p1, true);
  pentagonalNums.set(p2, true);

  while (d === 0 && p2 - p1 > d) {
    n++;
    p1 = p2;
    p2 = findPentagonal(n);
    pentagonalNums.set(p2, true);

    if (d === 0) {
      for (let n0 = n - 1; n0 >= 1; n0--) {
        const p0 = findPentagonal(n0);
        if (isPentagonal(p2 - p0, pentagonalNums) && isPentagonal(p2 + p0, pentagonalNums)) {
          d = p2 - p0;
          break;
        }
      }
    } else {
      let n0 = n - 1;
      let p0 = p1;
      while (p2 - p0 < d || n0 > 0) {
        if (isPentagonal(p2 - p0, pentagonalNums) && isPentagonal(p2 + p0, pentagonalNums)) {
          d = p2 - p0;
          break;
        }
        n0--;
        p0 = findPentagonal(n0);
      }
    }
  }

  return d;
}
